### ISSUE TITLE
Wrong PossiblyNullReference on switch case condition

### ISSUE THREAD

**Issue by gianiaz** [vimeo/psalm#3563]
Hi, I wrote a simple example:

```php
<?php

$a = rand(0,1) ? new \DateTime() : null;

switch(true) {
    case $a !== null && $a->format('Y') === '2020':
        $a->format('d-m-Y');
}
```
https://psalm.dev/r/b181d835cb

In this case psalm should not warn about a possibly null reference on format call because $a is already !== null.

I think the root cause is the condition in &&. If I remove the second condition psalm detects the $a is !== null.

**Comment by muglug** [comment]
This is (IMO) bad code that Psalm shouldn't understand.

It's equivalent to:

```
<?php

$a = rand(0,1) ? new DateTime() : null;

if (($a !== null && $a->format('Y') === '2020') == true) {
    $a->format('d-m-Y');
}
```

which, again, is bad code.

**Comment by gianiaz** [comment]
I'm sorry but I can't understand, It's only an example, even if it's bad code it's a valid code :-)

For example phpstan does not complain about this.

**Comment by Jean85** [comment]
Obviously the code was a simplified version of the original one where we stumbled on this issue, to make the bug reproduction easier.

The original code had multiple `case` instructions with similar repeated checks. For example, I tend personally to use `switch(true)` when I need multiple `instanceof` of the same class for example, because it makes it easier to spot the repeating pattern.

I don't consider it excellent code, but it's valid and merits a fix. If you can point us to the right test, we can try to write up at least a regression.

**Comment by muglug** [comment]
Ugh ok one second.
